src/kafka/utils/retry_helper.hh

```diff
@@ -0,0 +1,99 @@
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/*
+ * Copyright (C) 2019 ScyllaDB Ltd.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <cmath>
+#include <random>
+#include <chrono>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/sleep.hh>
+#include <seastar/util/bool_class.hh>
+
+namespace seastar {
+
+namespace kafka {
+
+struct do_retry_tag { };
+using do_retry = bool_class<do_retry_tag>;
+
+class retry_helper {
+private:
+    uint32_t _max_retry_count;
+    float _base_ms;
+    uint32_t _max_backoff_ms;
+
+    std::random_device _rd;
+    std::mt19937 _mt;
+
+    template<typename AsyncAction, typename DataType>
+    future<> with_retry(DataType& data, AsyncAction&& action, uint32_t retry_number) {
+        if (retry_number >= _max_retry_count) {
+            return make_ready_future<>();
+        }
+        return backoff(retry_number)
+        .then([this, &data, action = std::forward<AsyncAction>(action), retry_number]() mutable {
+            return futurize_apply(action, data)
+            .then([this, &data, action = std::forward<AsyncAction>(action), retry_number](auto do_retry_val) mutable {
+                if (do_retry_val == do_retry::yes) {
+                    return with_retry(data, std::forward<AsyncAction>(action), retry_number + 1);
+                } else {
+                    return make_ready_future<>();
+                }
+            });
+        });
+    }
+
+    future<> backoff(uint32_t retry_number) {
+        if (retry_number == 0) {
+            return make_ready_future<>();
+        }
+
+        // Exponential backoff with (full) jitter
+        auto backoff_time = _base_ms * std::pow(2.0f, retry_number - 1);
+        auto backoff_time_discrete = static_cast<uint32_t>(std::round(backoff_time));
+
+        auto capped_backoff_time = std::min(_max_backoff_ms, backoff_time_discrete);
+        std::uniform_int_distribution<uint32_t> dist(0, capped_backoff_time);
+
+        auto jittered_backoff = dist(_mt);
+        return seastar::sleep(std::chrono::milliseconds(jittered_backoff));
+    }
+
+public:
+    retry_helper(uint32_t max_retry_count, float base_ms, uint32_t max_backoff_ms)
+        : _max_retry_count(max_retry_count), _base_ms(base_ms), _max_backoff_ms(max_backoff_ms), _mt(_rd()) {}
+
+    template<typename AsyncAction, typename DataType>
+    future<> with_retry(DataType&& data, AsyncAction&& action) {
+        return do_with(std::forward<DataType>(data),
+                [this, action = std::forward<AsyncAction>(action)](auto& data_ref) mutable {
+            return with_retry(data_ref, std::forward<AsyncAction>(action), 0);
```

Comment:
This reference boilerplating seems to be a bit complicated. After some consideration, I'd suggest maybe leveraging the fact that `action` can be a **mutable** lambda and thus, once having captured something by value, it keeps it through every function call (and mutable lambdas keep the values modified respectively). Then, we can just move the `action` around.
In terms of code it would just mean that this can be one function `future<> with_retry(AsyncAction&& action)`. The action object itself holds the data relevant, so there's no need for us to take it as a parameter nor pass it anywhere ourselves. Whenever something needs to call `action`, just capture it as an rvalue ref.
Then, when using `with_retry` it would look something like this:
```cpp
DataType data {"Some fine lookin data chunk"};

// invoking without changing data between retries
(void*) helper.with_retry([data] {
    auto response = send(data);
    if (response.succeeded()) {
        return kafka::do_retry::no;
    }
    return kafka::do_retry::yes;
});

// invoking with modifying allowed
(void*) helper.with_retry([data] mutable {
    auto response = send(data);
    if (response.succeeded()) {
        return kafka::do_retry::no;
    }
    data += "I'm a failure"; // this modification persists between function calls
    return kafka::do_retry::yes;
});
```

Another thing is the `std::forward`. I'm not 100% sure whether it can deduce that it should actually move here and in the places above (although probably it's able to). If there is no specific reason for using this here, I'd rather advise using `action = std::move(action)` for safety.